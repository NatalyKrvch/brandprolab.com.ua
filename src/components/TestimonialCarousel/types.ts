import { TestimonialCardProps } from '../Cards/TestimonialCard/types';

export interface TestimonialCarouselProps {
  testimonials: TestimonialCardProps[];
}
