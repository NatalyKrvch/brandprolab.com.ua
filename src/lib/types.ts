export type Size = 's' | 'm' | 'l';
export type Screen = 'mobile' | 'tablet' | 'desktop';
