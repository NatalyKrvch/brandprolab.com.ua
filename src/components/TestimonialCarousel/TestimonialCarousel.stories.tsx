import type { Meta, StoryObj } from '@storybook/react';

import TestimonialCarousel from './TestimonialCarousel';
import type { TestimonialCarouselProps } from './types';

const meta: Meta<typeof TestimonialCarousel> = {
  title: 'Components/TestimonialCarousel',
  component: TestimonialCarousel,
  parameters: {
    controls: { expanded: true },
  },
  argTypes: {
    testimonials: {
      description:
        'An array of testimonial objects to display in the carousel.',
      table: {
        type: {
          summary: 'Array<Testimonial>',
        },
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof TestimonialCarousel>;

export const Default: Story = {
  args: {
    testimonials: [
      {
        text: 'The team was amazing to work with! Highly professional and dedicated.',
        clientName: 'Anna Kovalchuk',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/annakovalchuk',
      },
      {
        text: 'They exceeded our expectations with the final product.',
        clientName: 'Dmytro Shevchenko',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/dmytroshevchenko',
      },
      {
        text: 'Great communication and smooth collaboration from start to finish.',
        clientName: 'Olena Bondarenko',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/olenabondarenko',
      },
      {
        text: 'The team was amazing to work with! Highly professional and dedicated.',
        clientName: 'Anna Kovalchuk',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/annakovalchuk',
      },
      {
        text: 'They exceeded our expectations with the final product.',
        clientName: 'Dmytro Shevchenko',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/dmytroshevchenko',
      },
      {
        text: 'Great communication and smooth collaboration from start to finish.',
        clientName: 'Olena Bondarenko',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/olenabondarenko',
      },
      {
        text: 'The team was amazing to work with! Highly professional and dedicated.',
        clientName: 'Anna Kovalchuk',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/annakovalchuk',
      },
      {
        text: 'They exceeded our expectations with the final product.',
        clientName: 'Dmytro Shevchenko',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/dmytroshevchenko',
      },
      {
        text: 'Great communication and smooth collaboration from start to finish.',
        clientName: 'Olena Bondarenko',
        clientPhotoUrl:
          'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
        clientLink: 'https://linkedin.com/in/olenabondarenko',
      },
    ],
  } satisfies TestimonialCarouselProps,
};
