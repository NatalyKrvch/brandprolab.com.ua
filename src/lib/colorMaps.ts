import {
  MainSocialCardColor,
  socialMediaColorClassMapType,
} from '@/components/Cards/SocialMediaCard/types';

export const socialMediaColorClassMap: socialMediaColorClassMapType = {
  [MainSocialCardColor.RED]: {
    bg: 'bg-red-default',
    text: 'text-red-default',
  },
  [MainSocialCardColor.BLUE]: {
    bg: 'bg-blue-default',
    text: 'text-blue-default',
  },
};
