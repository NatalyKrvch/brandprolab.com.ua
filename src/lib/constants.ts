export const ALL_DATA_QUERY = `
{
  "header": *[_type == "header"][0],
  "hero": *[_type == "hero"][0],
  "cases": *[_type == "cases"][0],
  "services": *[_type == "services"][0],
  "testimonials": *[_type == "testimonialsSection"][0],
  "subscribe": *[_type == "subscribeSection"][0],
  "video": *[_type == "videoSection"][0],
  "expertise": *[_type == "expertiseSection"][0],
  "diagnostic": *[_type == "diagnosticSection"][0],
  "footer": *[_type == "footerSection"][0]
}
`;

export const YOUTUBE_REGEX =
  /(?:youtu\.be\/|youtube\.com\/(?:.*[?&]v=|embed\/|v\/|shorts\/|live\/))([^&?/]+)/;

export enum IconClassType {
  SERVICES = 'services',
  EXPERTISE = 'expertise',
  SOCIAL_MEDIA = 'social-media',
  SOCIAL_MEDIA_MAIN = 'social-media-main',
}
export const REVALIDATE = 60;
export const ENDLESS_CAROUSEL_SPEED = 50;
export const SLIDER_RIGHT_SHIFT = 2;
export const SLIDER_LEFT_SHIFT = -1;
export const AMOUNT_OF_TESTIMONIALS_WORDS = 15;
export const AMOUNT_OF_CASES_WORDS_DESKTOP = 13;
export const AMOUNT_OF_CASES_WORDS_TABLET = 9;
export const AMOUNT_OF_CASES_WORDS_MOBILE = 13;
