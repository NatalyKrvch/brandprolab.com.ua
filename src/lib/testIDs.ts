export const INTERNAL_LINK_TEST_ID = 'internal-link-test-id';
export const EXTERNAL_LINK_TEST_ID = 'external-link-test-id';
export const FULL_WIDTH_CONTAINER_TEST_ID = 'full-width-container-test-id';
export const LIMITED_WIDTH_CONTAINER_TEST_ID =
  'limited-width-container-test-id';
export const LEFT_LIMITED_CONTAINER_TEST_ID = 'left-limited-container-test-id';
export const ICON_COMPONENT_TEST_ID = 'icon-component-test-id';
export const ICON_COMPONENT_IMAGE_TEST_ID = 'icon-component-image-test-id';
export const LABEL_COMPONENT_TEST_ID = 'label-component-test-id';
export const BUTTON_TEST_ID = 'button-test-id';
export const UNDERLINED_TEXT_TEST_ID = 'underlined-text-test-id';
export const USER_BADGE_TEST_ID = 'user-badge-test-id';
export const USER_BADGE_IMAGE_TEST_ID = 'user-badge-image-test-id';
export const USER_BADGE_NAME_TEST_ID = 'user-badge-name-test-id';
export const TESTIMONIAL_CARD_TEST_ID = 'testimonials-card-test-id';
export const TESTIMONIAL_TEXT_TEST_ID = 'testimonials-text-test-id';
export const LIST_TEST_ID = 'list-test-id';
export const LIST_ITEM_TEST_ID = 'list-item-test-id';
export const SIMPLE_CARD_TEST_ID = 'simple-card-test-id';
export const DIPLOMA_CARD_TEST_ID = 'diploma-card-test-id';
export const DIPLOMA_IMAGE_TEST_ID = 'diploma-image-test-id';
export const DIPLOMA_OVERLAY_TEST_ID = 'diploma-overlay-test-id';
export const PHOTO_CARD_IMAGE_TEST_ID = 'photo-card-image-test-id';
export const PHOTO_CARD_TEXT_TEST_ID = 'photo-card-text-test-id';
export const VIDEO_IFRAME_TEST_ID = 'video-iframe-test-id';
export const VIDEO_PLAY_BUTTON_TEST_ID = 'video-play-button-test-id';
export const VIDEO_TEST_ID = 'video-test-id';
export const VIDEO_THUMBNAIL_TEST_ID = 'video-thumbnail-test-id';
export const FLIP_CARD_BACK_TEST_ID = 'flip-card-back-test-id';
export const FLIP_CARD_FRONT_TEST_ID = 'flip-card-front-test-id';
export const FLIP_CARD_TEST_ID = 'flip-card-test-id';
export const SERVICE_CARD_FRONT_DESCRIPTION_TEST_ID =
  'service-card-front-description-test-id';
export const SERVICE_CARD_FRONT_HEADER_TEST_ID =
  'service-card-front-header-test-id';
export const SERVICE_CARD_BACK_HEADER_TEST_ID =
  'service-card-back-header-test-id';
export const SERVICE_CARD_BACK_LIST_ITEM_TEST_ID =
  'service-card-back-list-item-test-id';
export const SERVICE_CARD_BACK_LIST_TEST_ID = 'service-card-back-list-test-id';
export const SOCIAL_MEDIA_LABEL_TEST_ID = 'social-media-label-test-id';
export const SOCIAL_MEDIA_DESCRIPTION_TEST_ID =
  'social-media-description-test-id';
export const FOOTER_SECTION_TEST_ID = 'footer-section-test-id';
export const PAGINATION_DOT_TEST_ID = 'pagination-dot-test-id';
export const CAROUSEL_NAV_WRAPPER_TEST_ID = 'carousel-nav-wrapper-test-id';
export const CAROUSEL_NAV_NEXT_TEST_ID = 'carousel-nav-next-test-id';
export const CAROUSEL_NAV_PREV_TEST_ID = 'carousel-nav-prev-test-id';
export const ENDLESS_CAROUSEL_TEST_ID = 'endless-carousel-test-id';
export const ENDLESS_CAROUSEL_CONTENT_TEST_ID =
  'endless-carousel-content-test-id';
export const MAIN_SPRAY_BG_TEST_ID = 'main-spray-bg-test-id';
export const HERO_MAIN_CARD_ICON_TEST_ID = 'hero-main-card-icon-test-id';
export const HERO_TEXT_CONTENT_TEST_ID = 'hero-text-content-test-id';
export const HERO_TEXT_CONTENT_ICON_TEST_ID = 'hero-text-content-icon-test-id';
export const EXPERTISE_SECTION_TEST_ID = 'expertise-section-test-id';
export const SCROLL_UP_BUTTON_TEST_ID = 'scroll-up-button-test-id';
export const MODAL_WRAPPER_TEST_ID = 'modal-wrapper-test-id';
export const MODAL_CONTENT_TEST_ID = 'modal-content-test-id';
export const MODAL_CLOSE_BUTTON_TEST_ID = 'modal-close-button-test-id';
export const CASE_MODAL_TEXT_TEST_ID = 'case-modal-text-test-id';
export const TESTIMONIAL_MODAL_NAME_TEST_ID = 'testimonial-modal-name-test-id';
export const TESTIMONIAL_MODAL_POSITION_TEST_ID =
  'testimonial-modal-position-test-id';
export const TESTIMONIAL_MODAL_TEXT_TEST_ID = 'testimonial-modal-text-test-id';
export const TESTIMONIAL_MODAL_IMAGE_TEST_ID =
  'testimonial-modal-image-test-id';
