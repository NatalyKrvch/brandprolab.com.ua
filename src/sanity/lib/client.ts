import { createClient } from 'next-sanity';

import { getSanityConfig } from '../env';

const { apiVersion, dataset, projectId, useCdn } = getSanityConfig();

export const client = createClient({
  projectId,
  dataset,
  apiVersion,
  useCdn,
});
