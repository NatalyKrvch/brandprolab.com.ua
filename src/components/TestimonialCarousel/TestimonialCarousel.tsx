import 'keen-slider/keen-slider.min.css';

import { useMemo } from 'react';

import { SLIDE_CLASSNAME, SLIDER_CLASSNAME } from '@/styles/constants';
import { generateRandomId } from '@/utils/generateRandomId';

import { TestimonialCard } from '../Cards/TestimonialCard';
import { CarouselNavButtons } from '../CarouselNavButtons';
import { PaginationDots } from '../PaginationDots';
import { useTestimonialCarousel } from './hooks/useTestimonialCarousel';
import { TestimonialCarouselProps } from './types';

const TestimonialCarousel = ({ testimonials }: TestimonialCarouselProps) => {
  const sliderId = useMemo(() => generateRandomId(), []);

  const {
    sliderRef,
    currentSlideIndex,
    isFirstSlide,
    isLastSlide,
    handleNextSlide,
    handlePreviousSlide,
    goToSlide,
  } = useTestimonialCarousel();

  return (
    <div className="relative">
      <CarouselNavButtons
        isFirstSlide={isFirstSlide}
        isLastSlide={isLastSlide}
        onPreviousSlide={handlePreviousSlide}
        onNextSlide={handleNextSlide}
      />

      <div ref={sliderRef} className={SLIDER_CLASSNAME}>
        {testimonials.map((testimonial, index) => (
          <div key={`${sliderId}-slide-${index}`} className={SLIDE_CLASSNAME}>
            <TestimonialCard {...testimonial} className="h-240 desktop:h-300" />
          </div>
        ))}
      </div>

      <PaginationDots
        total={testimonials.length}
        current={currentSlideIndex}
        onDotClick={goToSlide}
        className="mt-4 hidden mobile:flex"
      />
    </div>
  );
};

export default TestimonialCarousel;
