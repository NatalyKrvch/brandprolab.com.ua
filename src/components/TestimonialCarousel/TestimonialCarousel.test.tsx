import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import { TESTIMONIAL_CARD_TEST_ID } from '@/lib/testIDs';

import TestimonialCarousel from './TestimonialCarousel';

jest.mock('./hooks/useTestimonialCarousel', () => ({
  useTestimonialCarousel: () => ({
    sliderRef: jest.fn(),
    currentSlideIndex: 0,
    isFirstSlide: true,
    isLastSlide: false,
    handleNextSlide: jest.fn(),
    handlePreviousSlide: jest.fn(),
    goToSlide: jest.fn(),
  }),
}));

jest.mock('../Cards/TestimonialCard', () => ({
  TestimonialCard: ({ className }: { className?: string }) => (
    <div data-testid={TESTIMONIAL_CARD_TEST_ID} className={className}>
      Mocked Testimonial Card
    </div>
  ),
}));

jest.mock('../CarouselNavButtons', () => ({
  CarouselNavButtons: () => (
    <div data-testid="carousel-nav-buttons-mock">Mocked Nav Buttons</div>
  ),
}));

jest.mock('../PaginationDots', () => ({
  PaginationDots: () => (
    <div data-testid="pagination-dots-mock">Mocked Pagination Dots</div>
  ),
}));

describe('TestimonialCarousel component', () => {
  const mockTestimonials = [
    { text: 'Test 1', clientName: 'User 1', clientPhotoUrl: '/img1.jpg' },
    { text: 'Test 2', clientName: 'User 2', clientPhotoUrl: '/img2.jpg' },
    { text: 'Test 3', clientName: 'User 3', clientPhotoUrl: '/img3.jpg' },
  ];

  beforeAll(() => {
    Object.defineProperty(global, 'crypto', {
      value: {
        randomUUID: () => 'mocked-uuid',
      },
    });
  });

  it('renders all testimonial cards', () => {
    const { asFragment } = render(
      <TestimonialCarousel testimonials={mockTestimonials} />,
    );

    const testimonialCards = screen.getAllByTestId(TESTIMONIAL_CARD_TEST_ID);
    expect(testimonialCards).toHaveLength(mockTestimonials.length);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders carousel navigation buttons', () => {
    render(<TestimonialCarousel testimonials={mockTestimonials} />);
    const navButtons = screen.getByTestId('carousel-nav-buttons-mock');
    expect(navButtons).toBeInTheDocument();
  });

  it('renders pagination dots', () => {
    render(<TestimonialCarousel testimonials={mockTestimonials} />);
    const pagination = screen.getByTestId('pagination-dots-mock');
    expect(pagination).toBeInTheDocument();
  });
});
