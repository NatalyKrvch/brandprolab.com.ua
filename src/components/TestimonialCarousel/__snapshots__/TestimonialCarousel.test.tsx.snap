// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestimonialCarousel component renders all testimonial cards 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      data-testid="carousel-nav-buttons-mock"
    >
      Mocked Nav Buttons
    </div>
    <div
      class="keen-slider"
    >
      <div
        class="keen-slider__slide"
      >
        <div
          class="h-240 desktop:h-300"
          data-testid="testimonials-card-test-id"
        >
          Mocked Testimonial Card
        </div>
      </div>
      <div
        class="keen-slider__slide"
      >
        <div
          class="h-240 desktop:h-300"
          data-testid="testimonials-card-test-id"
        >
          Mocked Testimonial Card
        </div>
      </div>
      <div
        class="keen-slider__slide"
      >
        <div
          class="h-240 desktop:h-300"
          data-testid="testimonials-card-test-id"
        >
          Mocked Testimonial Card
        </div>
      </div>
    </div>
    <div
      data-testid="pagination-dots-mock"
    >
      Mocked Pagination Dots
    </div>
  </div>
</DocumentFragment>
`;
